WordPress/Sniffs/WP/TemplatedInlineScriptsSniff.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ * WordPress Coding Standard.
+ *
+ * @package WPCS\WordPressCodingStandards
+ * @link    https://github.com/WordPress/WordPress-Coding-Standards
+ * @license https://opensource.org/licenses/MIT MIT
+ */
+
+namespace WordPressCS\WordPress\Sniffs\WP;
+
+use PHP_CodeSniffer\Exceptions\RuntimeException;
+use PHP_CodeSniffer\Util\Tokens;
+use PHPCSUtils\Tokens\Collections;
+use PHPCSUtils\Utils\TextStrings;
+use WordPressCS\WordPress\Sniff;
+
+/**
+ * Makes sure inline scripts and scripts are generated using either {@see wp_get_script_tag()} or {@see wp_get_inline_script_tag()}.
+ *
+ * @package WPCS\WordPressCodingStandards
+ * @since   3.0.0
+ */
+class TemplatedInlineScriptsSniff extends Sniff {
+
+	/**
+	 * Returns an array of tokens this test wants to listen for.
+	 *
+	 * @return array
+	 */
+	public function register() {
+		$targets   = Collections::$textStingStartTokens;
+		$targets[] = \T_INLINE_HTML;
+
+		return $targets;
+	}
+
+	/**
+	 * Processes this test, when one of its tokens is encountered.
+	 *
+	 * @param int $stackPtr The position of the current token in the stack.
+	 *
+	 * @return void
+	 */
+	public function process_token( $stackPtr ) {
+
+		$content = $this->tokens[ $stackPtr ]['content'];
+		if ( \T_INLINE_HTML !== $this->tokens[ $stackPtr ]['code'] ) {
+			try {
+				$content = TextStrings::getCompleteTextString( $this->phpcsFile, $stackPtr );
+			} catch ( RuntimeException $e ) {
+				// Not the first token in a multi-line text string. Any issues will already have been reported.
+				return;
+			}
+		}
+
```

Comment:
Add a comment that `<script src>` is already handled by `EnqueuedResources`.

Reply:
> Hi @enricocarraro, welcome to this repo! Thanks for the PR.
> 
> Checklist of things to do before this PR will be properly reviewed:
> 
> * [ ]  Add unit tests.
>   Please make sure they also contain good code samples testing against false positives and false negatives.
> * [ ]  Add XML documentation for the sniff to the WordPress/Docs directory.
> * [ ]  The build needs to pass.
> 
> Please feel free to ask for as much guidance as you need from me and the others in the team.
> 
> In my case, my remarks will often presume you know what you're doing with PHPCS (as why would you send in a sniff otherwise?), please don't let that intimidate you and ask for as much clarification as you need.
> 

> 

> 
> Other notes:
> 
> * The (complex) regex should probably be a class constant and should be properly documented.

> * We're not a fan of duplicate code and most of this sniff seems to be copied from another sniff.
>   



Hi @jrfnl, thanks for taking the time to look into this PR and pointing out what is not clear about it.

> I've had a quick look at it so far and to me it's unclear what this sniff is looking for based on the PR description, so I can't review the code to see if the sniff actually does what it is supposed to do.
The sniff is looking for script tags that contain inline JavaScript, since the `EnqueuedResources` sniff only looks for script loaders tags (script tags with the `src` attribute).

> * Is `WP` the right category for this sniff or should this sniff be in the `Security` category ? I honestly don't know at the moment, but would like to see this discussed.
I am not sure what the most suited category would be, I placed it into `WP` since as you mentioned, `TemplatedScripts` is similar to `EnqueuedResources`. In WordPress 5.7 [enqueued scripts will be printed using the new functions](https://github.com/WordPress/wordpress-develop/pull/498/files#diff-260c63dc48da5ecd590768aadf946b47658284c7b86035c0aedd3dbfee2b20b5) that enable strict CSP-compatibility.

> Aside from that, the PR description refers to a function which won't be introduced in WP until version `5.7`. That implies that the sniff will need to support the `$minimum_supported_version` property, as WPCS is used by both Core, as well as plugins and themes which often support multiple WP versions.
I will make sure to include the property in the next commits.

>   1. Should this check be added to that other sniff instead ?
In retrospect I think it would be better to add this new sniff under `EnqueuedResourcesSniff`.

Reply:
Hi @jrfnl, thanks for taking the time to look into this PR and pointing out what is not clear about it.

> I've had a quick look at it so far and to me it's unclear what this sniff is looking for based on the PR description, so I can't review the code to see if the sniff actually does what it is supposed to do.

The sniff is looking for script tags that contain inline JavaScript, since the `EnqueuedResources` sniff only looks for script loaders tags (script tags with the `src` attribute).

> * Is `WP` the right category for this sniff or should this sniff be in the `Security` category ? I honestly don't know at the moment, but would like to see this discussed.

I am not sure what the most appropriate category would be; I placed it into `WP` since, as you mentioned, `TemplatedScripts` is similar to `EnqueuedResources`. In WordPress 5.7 [enqueued scripts will be printed using the new functions](https://github.com/WordPress/wordpress-develop/pull/498/files#diff-260c63dc48da5ecd590768aadf946b47658284c7b86035c0aedd3dbfee2b20b5) that enable strict CSP-compatibility.

> Aside from that, the PR description refers to a function which won't be introduced in WP until version `5.7`. That implies that the sniff will need to support the `$minimum_supported_version` property, as WPCS is used by both Core, as well as plugins and themes which often support multiple WP versions.

I will make sure to include the property in the next commits.

>   1. Should this check be added to that other sniff instead ?

In retrospect, I think it would be better to add this new sniff under `EnqueuedResourcesSniff`.

> The sniff has now been added to Extra. As the description mentions fixing this in WP Core, should the sniff be part of the Core ruleset ?

In my opinion, both `NonEnqueuedScript` and `NonTemplatedInlineScript` should be moved to the Core ruleset to prevent strict CSP-compatibility backsliding effectively starting from WordPress 5.7.